src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalescingAssignmentTests.cs

```diff
@@ -0,0 +1,1408 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.UnitTests;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests.Semantics
+{
+    [CompilerTrait(CompilerFeature.NullCoalescingAssignment)]
+    public class NullCoalescingAssignmentTests : CompilingTestBase
+    {
+        [Fact]
+        public void LocalLvalue()
+        {
+
+            CompileAndVerify(@"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        int? i1 = null;
+        Console.WriteLine(i1 ??= GetInt());
+    }
+    static int GetInt()
+    {
+        Console.WriteLine(""In GetInt"");
+        return 0;
+    }
+}
+", expectedOutput: @"
+In GetInt
+0
+").VerifyIL("C.Main()", @"
+{
+  // Code size       43 (0x2b)
+  .maxstack  1
+  .locals init (int? V_0)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    ""int?""
+  IL_0008:  ldloc.0
+  IL_0009:  stloc.0
+  IL_000a:  ldloca.s   V_0
+  IL_000c:  call       ""bool int?.HasValue.get""
+  IL_0011:  brtrue.s   IL_001f
+  IL_0013:  call       ""int C.GetInt()""
+  IL_0018:  newobj     ""int?..ctor(int)""
+  IL_001d:  br.s       IL_0020
+  IL_001f:  ldloc.0
+  IL_0020:  box        ""int?""
+  IL_0025:  call       ""void System.Console.WriteLine(object)""
+  IL_002a:  ret
+}
+");
+        }
+
+        [Fact]
+        public void FieldLvalue()
+        {
+
+            CompileAndVerify(@"
+using System;
+public class C
+{
+    int? f1 = null;
+
+    public static void Main()
+    {
+        var c = new C();
+        Console.WriteLine(c.f1 ??= GetInt());
+    }
+    static int GetInt()
+    {
+        Console.WriteLine(""In GetInt"");
+        return 0;
+    }
+}
+", expectedOutput: @"
+In GetInt
+0
+").VerifyIL("C.Main()", @"
+{
+  // Code size       55 (0x37)
+  .maxstack  3
+  .locals init (C V_0,
+                int? V_1,
+                int? V_2)
+  IL_0000:  newobj     ""C..ctor()""
+  IL_0005:  stloc.0
+  IL_0006:  ldloc.0
+  IL_0007:  ldfld      ""int? C.f1""
+  IL_000c:  stloc.1
+  IL_000d:  ldloca.s   V_1
+  IL_000f:  call       ""bool int?.HasValue.get""
+  IL_0014:  brtrue.s   IL_002b
+  IL_0016:  ldloc.0
+  IL_0017:  call       ""int C.GetInt()""
+  IL_001c:  newobj     ""int?..ctor(int)""
+  IL_0021:  dup
+  IL_0022:  stloc.2
+  IL_0023:  stfld      ""int? C.f1""
+  IL_0028:  ldloc.2
+  IL_0029:  br.s       IL_002c
+  IL_002b:  ldloc.1
+  IL_002c:  box        ""int?""
+  IL_0031:  call       ""void System.Console.WriteLine(object)""
+  IL_0036:  ret
+}
+");
+        }
```

Comment:
>} [](start = 8, length = 1)

If you don't have that already, can you test with LHS being a struct type (ie. could not be null), unconstrained type `T` and constrained type `T`?
Also, various typeless expressions (`null`, `default`, lambdas, typeless tuple `(null, null)`).

Reply:
`InvalidLHS` has a value-type LHS, and `TypeParameterLHS` has constrained and unconstrained `T`. `null` and lambdas were also tested, I'll expand the tuples and default tests in a followup.

---
In reply to: [200780305](https://github.com/dotnet/roslyn/pull/28253#discussion_r200780305) [](ancestors = 200780305)